### PR TITLE
Add check for connection status before sync

### LINKED ISSUE #7526: Cancel Database Analyze & Sync and Add DB connection check before start
### Bug / Feature Request

- Your browser and the version: 66.0.3359.139
- Your operating system: OS X High Sierra
- Your databases: Redshift
- Metabase version: 0.28.6
- Metabase hosting environment: EC2
- Metabase internal database: Postgres

Steps to reproduce the issue
- Add a connection to a huge database
- Wait for it to sync / analyze
- Exclude the database (make it non existent)
- Trigger sync and analyze
- Metabase will try to resync each table/column even if the connection is not working

I think there should be a way to cancel the process and metabase should check the connection. If it is not working it shouldn't even start.

```
Unable to determine row count for Table 9852 'whatever_table.whatever_field': class java.lang.NullPointerException
("db.metadata_queries$table_row_count$fn__40646.invoke(metadata_queries.clj:39)"
"db.metadata_queries$table_row_count.invokeStatic(metadata_queries.clj:39)"
"db.metadata_queries$table_row_count.invoke(metadata_queries.clj:32)"
"sync.analyze.table_row_count$fn__40665$table_row_count__40670$fn__40671$fn__40672.invoke(table_row_count.clj:17)"
"sync.util$do_with_error_handling.invokeStatic(util.clj:124)"
"sync.util$do_with_error_handling.invoke(util.clj:119)"
"sync.analyze.table_row_count$fn__40665$table_row_count__40670$fn__40671.invoke(table_row_count.clj:16)"
"sync.analyze.table_row_count$fn__40665$table_row_count__40670.invoke(table_row_count.clj:13)"
"sync.analyze.table_row_count$fn__40690$update_row_count_BANG___40695$fn__40696$fn__40697.invoke(table_row_count.clj:23)"
"sync.util$do_with_error_handling.invokeStatic(util.clj:124)"
"sync.util$do_with_error_handling.invoke(util.clj:119)"
"sync.analyze.table_row_count$fn__40690$update_row_count_BANG___40695$fn__40696.invoke(table_row_count.clj:22)"
"sync.analyze.table_row_count$fn__40690$update_row_count_BANG___40695.invoke(table_row_count.clj:19)"
"sync.analyze$fn__40742$analyze_table_BANG___40747$fn__40748.invoke(analyze.clj:71)"
"sync.analyze$fn__40742$analyze_table_BANG___40747.invoke(analyze.clj:68)"
"sync.analyze$fn__40765$analyze_db_BANG___40770$fn__40771$fn__40772.invoke(analyze.clj:86)"
"sync.util$do_with_error_handling.invokeStatic(util.clj:124)"
"sync.util$do_with_error_handling.invoke(util.clj:119)"
"sync.util$do_with_error_handling.invokeStatic(util.clj:122)"
"sync.util$do_with_error_handling.invoke(util.clj:119)"
"driver$fn__26195.invokeStatic(driver.clj:235)"
"driver$fn__26195.invoke(driver.clj:235)"
"driver$fn__26130$G__25937__26139.invoke(driver.clj:51)"
"sync.util$sync_in_context$fn__39411.invoke(util.clj:115)"
"sync.util$with_db_logging_disabled$fn__39408.invoke(util.clj:106)"
"sync.util$with_start_and_finish_logging$fn__39403.invoke(util.clj:92)"
"sync.util$with_sync_events$fn__39400.invoke(util.clj:75)"
"sync.util$with_duplicate_ops_prevented$fn__39391.invoke(util.clj:54)"
"sync.util$do_sync_operation.invokeStatic(util.clj:142)"
"sync.util$do_sync_operation.invoke(util.clj:139)"
"sync.analyze$fn__40765$analyze_db_BANG___40770$fn__40771.invoke(analyze.clj:82)"
"sync.analyze$fn__40765$analyze_db_BANG___40770.invoke(analyze.clj:77)"
"task.sync_databases.SyncAndAnalyzeDatabase.execute(sync_databases.clj:39)")
```

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### PR DESCRIPTION
**Pull request by calherries**
Fixes https://github.com/metabase/metabase/issues/7526, except for oracle, vertica, and presto-jdbc.

It also fixes a few bugs in the driver `test.data` code.

Adds a check for valid connections in the following places: 
1. when handling a request to `POST /api/database/:id/sync_schema` (i.e. a manual trigger), we return a 422 status code
2. when a scheduled sync is triggered, we log a warning

In both cases if the connection fails we don't start any of the sync steps.

### Example with MySQL

**before a POST /api/database/:id/sync_schema with a deleted database would yield logs like this:**
```
2023-11-21 17:29:24,073 INFO metabase.events :: Publishing :event/database-manual-sync event (name and id):

{}

2023-11-21 17:29:24,093 DEBUG middleware.log :: POST /api/database/427/sync_schema 200 53.6 ms (7 DB calls) App DB connections: 0/10 Jetty threads: 5/50 (2 idle, 0 queued) (148 total active threads) Queries in flight: 0 (0 queued); mysql DB 427 connections: 0/1 (0 threads blocked)
2023-11-21 17:29:24,130 INFO metabase.events :: Publishing :event/sync-metadata-begin event (name and id):

{}

2023-11-21 17:29:24,130 INFO sync.util :: STARTING: Sync metadata for mysql Database 427 'foo'
2023-11-21 17:29:24,131 INFO sync.util :: STARTING: step 'sync-dbms-version' for mysql Database 427 'foo'
2023-11-21 17:29:24,133 INFO sync.util :: FINISHED: step 'sync-dbms-version' for mysql Database 427 'foo' (2.6 ms)
2023-11-21 17:29:24,134 INFO sync.util :: STARTING: step 'sync-timezone' for mysql Database 427 'foo'
2023-11-21 17:29:24,167 INFO sync-metadata.sync-timezone :: :mysql database 427 default timezone is "EET"
2023-11-21 17:29:24,167 INFO sync.util :: FINISHED: step 'sync-timezone' for mysql Database 427 'foo' (33.1 ms)
2023-11-21 17:29:24,167 INFO sync.util :: STARTING: step 'sync-tables' for mysql Database 427 'foo'
2023-11-21 17:29:24,170 INFO sync.util :: FINISHED: step 'sync-tables' for mysql Database 427 'foo' (3.2 ms)
2023-11-21 17:29:24,171 INFO sync.util :: STARTING: step 'sync-fields' for mysql Database 427 'foo'
2023-11-21 17:29:24,177 INFO sync.util :: FINISHED: step 'sync-fields' for mysql Database 427 'foo' (5.7 ms)
2023-11-21 17:29:24,177 INFO sync.util :: STARTING: step 'sync-fks' for mysql Database 427 'foo'
2023-11-21 17:29:24,178 INFO sync.util :: FINISHED: step 'sync-fks' for mysql Database 427 'foo' (1.2 ms)
2023-11-21 17:29:24,178 INFO sync.util :: STARTING: step 'sync-metabase-metadata' for mysql Database 427 'foo'
2023-11-21 17:29:24,178 INFO sync.util :: FINISHED: step 'sync-metabase-metadata' for mysql Database 427 'foo' (118.3 µs)
2023-11-21 17:29:24,178 INFO sync.util :: STARTING: step 'sync-table-privileges' for mysql Database 427 'foo'
2023-11-21 17:29:24,187 WARN sync.util :: Error running step 'sync-table-privileges' for mysql Database 427 'foo'
java.sql.SQLSyntaxErrorException: (conn=284871) Unknown database 'foo'
        at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.createException(ExceptionFactory.java:62)
        at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.create(ExceptionFactory.java:158)
        at org.mariadb.jdbc.MariaDbStatement.executeExceptionEpilogue(MariaDbStatement.java:262)
        at org.mariadb.jdbc.ClientSidePreparedStatement.executeInternal(ClientSidePreparedStatement.java:229)
        at org.mariadb.jdbc.ClientSidePreparedStatement.execute(ClientSidePreparedStatement.java:149)
        at org.mariadb.jdbc.ClientSidePreparedStatement.executeQuery(ClientSidePreparedStatement.java:163)
        at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeQuery(NewProxyPreparedStatement.java:1471)
        at clojure.java.jdbc$execute_query_with_params.invokeStatic(jdbc.clj:1090)
        at clojure.java.jdbc$execute_query_with_params.invoke(jdbc.clj:1084)
        at clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1113)
        at clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:1093)
        at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1182)
        at clojure.java.jdbc$query.invoke(jdbc.clj:1144)
        at metabase.driver.mysql$eval114017$fn__114018.invoke(mysql.clj:871)
        at clojure.lang.MultiFn.invoke(MultiFn.java:234)
        at metabase.sync.sync_metadata.sync_table_privileges$sync_table_privileges_BANG_$fn__117231.invoke(sync_table_privileges.clj:18)
        at metabase.sync.sync_metadata.sync_table_privileges$sync_table_privileges_BANG_.invokeStatic(sync_table_privileges.clj:11)
        at metabase.sync.sync_metadata.sync_table_privileges$sync_table_privileges_BANG_.invoke(sync_table_privileges.clj:11)
        at clojure.lang.AFn.applyToHelper(AFn.java:154)
        at clojure.lang.AFn.applyTo(AFn.java:144)
        at clojure.core$apply.invokeStatic(core.clj:669)
        at clojure.core$apply.invoke(core.clj:662)
        at metabase.sync.util$run_step_with_metadata$fn__83649$fn__83651.doInvoke(util.clj:425)
        at clojure.lang.RestFn.invoke(RestFn.java:397)
        at metabase.sync.util$with_start_and_finish_logging_STAR_.invokeStatic(util.clj:132)
        at metabase.sync.util$with_start_and_finish_logging_STAR_.invoke(util.clj:126)
        at metabase.sync.util$with_start_and_finish_debug_logging.invokeStatic(util.clj:150)
        at metabase.sync.util$with_start_and_finish_debug_logging.invoke(util.clj:146)
        at metabase.sync.util$run_step_with_metadata$fn__83649.invoke(util.clj:420)
        at metabase.sync.util$run_step_with_metadata.invokeStatic(util.clj:415)
        at metabase.sync.util$run_step_with_metadata.invoke(util.clj:415)
        at metabase.sync.util$run_sync_operation$fn__83732$fn__83740.invoke(util.clj:531)
        at metabase.sync.util$run_sync_operation$fn__83732.invoke(util.clj:529)
        at metabase.sync.util$run_sync_operation.invokeStatic(util.clj:523)
        at metabase.sync.util$run_sync_operation.invoke(util.clj:523)
        at metabase.sync.sync_metadata$sync_db_metadata_BANG_$fn__117510$fn__117511.invoke(sync_metadata.clj:63)
        at metabase.sync.util$do_with_error_handling.invokeStatic(util.clj:191)
        at metabase.sync.util$do_with_error_handling.invoke(util.clj:184)
        at clojure.core$partial$fn__5910.invoke(core.clj:2647)
        at metabase.driver$eval65925$fn__65926.invoke(driver.clj:725)
        at clojure.lang.MultiFn.invoke(MultiFn.java:239)
        at metabase.sync.util$sync_in_context$fn__83563.invoke(util.clj:167)
        at metabase.sync.util$with_db_logging_disabled$fn__83560.invoke(util.clj:159)
        at metabase.sync.util$with_start_and_finish_logging_STAR_.invokeStatic(util.clj:132)
        at metabase.sync.util$with_start_and_finish_logging_STAR_.invoke(util.clj:126)
        at metabase.sync.util$with_start_and_finish_logging$fn__83547.invoke(util.clj:144)
        at metabase.sync.util$with_sync_events$fn__83537$fn__83541.invoke(util.clj:118)
        at metabase.sync.util$with_duplicate_ops_prevented$fn__83520.invoke(util.clj:90)
        at metabase.sync.util$do_sync_operation$fn__83582.invoke(util.clj:216)
        at metabase.sync.util$do_sync_operation.invokeStatic(util.clj:210)
        at metabase.sync.util$do_sync_operation.invoke(util.clj:210)
        at metabase.sync.sync_metadata$sync_db_metadata_BANG_$fn__117510.invoke(sync_metadata.clj:62)
        at metabase.sync.sync_metadata$sync_db_metadata_BANG_.invokeStatic(sync_metadata.clj:58)
        at metabase.sync.sync_metadata$sync_db_metadata_BANG_.invoke(sync_metadata.clj:58)
        at metabase.api.database$fn__210643$fn__210644.invoke(database.clj:998)
        at clojure.core$binding_conveyor_fn$fn__5823.invoke(core.clj:2047)
        at clojure.lang.AFn.call(AFn.java:18)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.mariadb.jdbc.internal.util.exceptions.MariaDbSqlException: Unknown database 'foo'
        at org.mariadb.jdbc.internal.util.exceptions.MariaDbSqlException.of(MariaDbSqlException.java:34)
        at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.exceptionWithQuery(AbstractQueryProtocol.java:195)
        at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.exceptionWithQuery(AbstractQueryProtocol.java:178)
        at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.executeQuery(AbstractQueryProtocol.java:322)
        at org.mariadb.jdbc.ClientSidePreparedStatement.executeInternal(ClientSidePreparedStatement.java:220)
        ... 57 more
Caused by: java.sql.SQLException: Unknown database 'foo'
        at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.readErrorPacket(AbstractQueryProtocol.java:1693)
        at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.readPacket(AbstractQueryProtocol.java:1555)
        at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.getResult(AbstractQueryProtocol.java:1518)
        at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.executeQuery(AbstractQueryProtocol.java:319)
        ... 58 more
2023-11-21 17:29:24,188 INFO sync.util :: FINISHED: step 'sync-table-privileges' for mysql Database 427 'foo' (9.9 ms)
2023-11-21 17:29:24,214 INFO sync.util :: FINISHED: Sync metadata for mysql Database 427 'foo' (83.8 ms)
2023-11-21 17:29:24,214 INFO metabase.events :: Publishing :event/sync-metadata-end event (name and id):

{}

2023-11-21 17:29:24,215 INFO metabase.events :: Publishing :event/analyze-begin event (name and id):

{}

2023-11-21 17:29:24,215 INFO sync.util :: STARTING: Analyze data for mysql Database 427 'foo'
2023-11-21 17:29:24,216 INFO sync.util :: STARTING: step 'fingerprint-fields' for mysql Database 427 'foo'
2023-11-21 17:29:24,217 INFO sync.util :: FINISHED: step 'fingerprint-fields' for mysql Database 427 'foo' (1.5 ms)
2023-11-21 17:29:24,217 INFO sync.util :: STARTING: step 'classify-fields' for mysql Database 427 'foo'
2023-11-21 17:29:24,217 INFO sync.util :: FINISHED: step 'classify-fields' for mysql Database 427 'foo' (51.1 µs)
2023-11-21 17:29:24,218 INFO sync.util :: STARTING: step 'classify-tables' for mysql Database 427 'foo'
2023-11-21 17:29:24,218 INFO sync.util :: FINISHED: step 'classify-tables' for mysql Database 427 'foo' (83.0 µs)
2023-11-21 17:29:24,221 INFO sync.util :: FINISHED: Analyze data for mysql Database 427 'foo' (5.8 ms)
2023-11-21 17:29:24,221 INFO metabase.events :: Publishing :event/analyze-end event (name and id):

{}
```

**after these changes:**
```
2023-11-21 17:31:07,973 WARN middleware.log :: POST /api/database/427/sync_schema 422 94.1 ms (7 DB calls) 
"Could not connect to address=(host=localhost)(port=3306)(type=master) : (conn=284877) Unknown database 'foo'"
```

### CODE REVIEW

File: src/metabase/api/database.clj [hunk 1]
```diff
@@ -983,10 +984,20 @@
   ;; just wrap this in a future so it happens async
   (let [db (api/write-check (t2/select-one Database :id id))]
     (events/publish-event! :event/database-manual-sync {:object db :user-id api/*current-user-id*})
-    (future
-      (sync-metadata/sync-db-metadata! db)
-      (analyze/analyze-db! db)))
-  {:status :ok})
+    (if-let [ex (try (do
```
Comment (qnkhuat) [inline src/metabase/api/database.clj:987]: this do is redundant

File: src/metabase/api/database.clj [hunk 2]
```diff
@@ -983,10 +984,20 @@
...
+                       ;; it's okay to allow testing H2 connections during sync. We only want to disallow you from testing them for the
+                       ;; purposes of creating a new H2 database.
+                       (binding [h2/*allow-testing-h2-connections* true]
```
Comment (qnkhuat) [inline src/metabase/api/database.clj:990]: is this binding only used in tests?

if so you can bind this directly when you call the API because `mt/user-http-request` now makes an in-thread request.

File: test/metabase/driver_test.clj [hunk 3]
```diff
@@ -68,6 +75,48 @@
                        :tables
                        (some :schema))))))))
 
+(deftest check-can-connect-before-sync-test
+  (testing "Database sync should short-circuit and fail if the database connection is not available"
+    (mt/test-drivers (mt/normal-drivers)
+      (let [;; create a database with a table and sync
+            database-name      (name (gensym))
```
Comment (qnkhuat) [inline test/metabase/driver_test.clj:82]:
```suggestion
            database-name      (mt/random-name)
```
you can use this instead

File: test/metabase/driver_test.clj [hunk 4]
```diff
@@ -68,6 +75,48 @@
...
+        (testing "after deleting a database, sync should fail"
+          (testing "1: sync-and-analyze-database! should log a warning and fail early"
+            (is (some
+                 log-match?
+                 (mt/with-log-messages-for-level :warn
+                   (#'task.sync-databases/sync-and-analyze-database!* (u/the-id db))))))
+          (testing "2: triggering the sync via the POST /api/database/:id/sync_schema endpoint should fail"
+            (mt/user-http-request :crowberto :post 422 (str "/database/" (u/the-id db) "/sync_schema"))))
+        ;; clean up the database
+        (t2/delete! :model/Database (u/the-id db))))))
```
Comment (qnkhuat) [inline test/metabase/driver_test.clj:118]: use `mt/with-model-cleanup` instead?

File: src/metabase/api/database.clj (hunk 2 above)
Comment (calherries) [reply src/metabase/api/database.clj:990]: No, "testing" in `h2/*allow-testing-h2-connections*` is not testing as in tests, it is testing as in checking the connection. This binding is used in many places outside tests  e.g. https://github.com/metabase/metabase/blob/cb2ba50f3d4a7b7fc9f1634b27681514b561d940/src/metabase/api/table.clj#L84

File: test/metabase/driver_test.clj (hunk 4 above)
Comment (calherries) [reply test/metabase/driver_test.clj:118]: `mt/with-model-cleanup` doesn't execute the toucan hooks, which is needed here to clean up connection resources

https://github.com/metabase/metabase/blob/cb2ba50f3d4a7b7fc9f1634b27681514b561d940/test/metabase/test/util.clj#L663-L664

File: src/metabase/api/database.clj (hunk 1 above)
Comment (calherries) [reply src/metabase/api/database.clj:987]: Fixed c8d2c626d9f3bbd551d059ebbdd87a3557ae48fe

File: modules/drivers/bigquery-cloud-sdk/test/metabase/test/data/bigquery_cloud_sdk.clj [hunk 5]
```diff
@@ -47,15 +47,17 @@
...
-      ;; All databases created during test runs by this JVM instance get a suffix based on the timestamp from when
-      ;; this namespace was loaded. This dataset will not be deleted after this test run finishes, since there is no
-      ;; reasonable hook to do so (from this test extension namespace), so instead we will rely on each run cleaning
-      ;; up outdated, transient datasets via the `transient-dataset-outdated?` mechanism.
-      :db    (str "v3_" s "__transient_" ns-load-time)
-      :table s)))
+  [identifier]
+  (str/replace (name identifier) "-" "_"))
+
+(defn test-dataset-id
```
Comment (calherries) [inline modules/drivers/bigquery-cloud-sdk/test/metabase/test/data/bigquery_cloud_sdk.clj:53]: I made a refactoring, changing `(normalize-name :db database-name)` to `(test-dataset-id database-name)`. 

The reason behind this change is there was a bug [here](https://github.com/metabase/metabase/pull/35964/files#diff-02e0ddde9d998d54b97680780db70f4ee393bcb76c6a4ecce56c9693ca947236R345) where a raw database-name was passed to `destroy-dataset!`. I believe that bug was caused by confusion around `database-name`, `dataset-name`, and `dataset-id`, all being used to refer to the same type of thing. With this refactoring I wanted to make it clearer that `destroy-dataset!` should always take a test `dataset-id`, which is distinct from `database-name`, which is a key on the db definition.

File: modules/drivers/bigquery-cloud-sdk/test/metabase/test/data/bigquery_cloud_sdk.clj [hunk 6]
```diff
@@ -320,27 +322,27 @@
...
        (catch Throwable e
-         (log/error (u/format-color 'red  "Failed to load BigQuery dataset %s." (pr-str database-name)))
+         (log/error (u/format-color 'red  "Failed to load BigQuery dataset %s." (pr-str dataset-id)))
          (log/error (u/pprint-to-str 'red (Throwable->map e)))
          (throw e))))))
 
 (defmethod tx/destroy-db! :bigquery-cloud-sdk
   [_ {:keys [database-name]}]
-  (destroy-dataset! database-name))
+  (destroy-dataset! (test-dataset-id database-name)))
```
Comment (calherries) [inline modules/drivers/bigquery-cloud-sdk/test/metabase/test/data/bigquery_cloud_sdk.clj:345]: This was a bug. `tx/destroy-db!` never actually worked.

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj [hunk 7]
```diff
@@ -225,13 +229,14 @@
...
-       (try
-         (binding [*mongo-connection* db]
-           (f *mongo-connection*))
-         (finally
-           (mg/disconnect mongo-client)
-           (log/debug (u/format-color 'cyan (trs "Closed MongoDB connection.")))))))))
+        (log/debug (u/format-color 'cyan (trs "Opened new MongoDB connection.")))
+        (try
+          (binding [*mongo-connection* db
+                    *mongo-client*     mongo-client]
```
Comment (calherries) [inline modules/drivers/mongo/src/metabase/driver/mongo/util.clj:235]: I added this to dynamic scope to make it available up [here](https://github.com/metabase/metabase/pull/35964/files#diff-e7783632d76b787db2581115d762107f65336b44af25298042df2da4ac1a7731R60). Call me lazy, but I didn't want to break the macro interface by making this an argument. I think it's worth it.

File: modules/drivers/mongo/test/metabase/driver/mongo_test.clj [hunk 8]
```diff
@@ -46,40 +46,42 @@
...
-                                                :expected (not (tdm/ssl-required?))}
-                                               {:details  {:conn-uri "mongodb://localhost:3000/bad-db-name?connectTimeoutMS=50"}
-                                                :expected false}]
-           :let [ssl-details (tdm/conn-details details)]]
-     (testing (str "connect with " details)
-       (is (= expected
-              (driver.u/can-connect-with-details? :mongo ssl-details))
-           (str message))))))
+   (mt/dataset sample-dataset
+     (mt/db)
```
Comment (calherries) [inline modules/drivers/mongo/test/metabase/driver/mongo_test.clj:50]: This test was effectively broken. It should have been failing but just happened to work. `dbname` was actually never checked to see if it was there or not, now it is checked, and I've updated the test to use the sample-dataset.

File: src/metabase/task/sync_databases.clj [hunk 9]
```diff
@@ -72,6 +74,30 @@
          end-time
          (< (.toMinutes (t/duration start-time end-time)) analyze-duration-threshold-for-refingerprinting))))
 
+(defn- sync-and-analyze-database*!
```
Comment (calherries) [inline src/metabase/task/sync_databases.clj:77]: I extracted the bulk of this out from `sync-and-analyze-database!` so we could test it.

File: test/metabase/driver_test.clj [hunk 10]
```diff
@@ -68,24 +72,74 @@
...
                        (some :schema))))))))
 
+(defn- basic-db-definition [database-name]
+  (tx/map->DatabaseDefinition
+   {:database-name     database-name
+    :table-definitions [{:table-name        "baz"
+                         :field-definitions [{:field-name "foo", :base-type :type/Text}]
+                         :rows              [["bar"]]}]}))
+
+(deftest check-can-connect-before-sync-test
```
Comment (calherries) [inline test/metabase/driver_test.clj:82]: Here's the big test.

File: test/metabase/driver_test.clj [hunk 11]
```diff
@@ -68,24 +72,75 @@
...
+                                          (re-matches #"^Cannot sync Database (.+): (.+)" message)))
+                                   (mt/with-log-messages-for-level :warn
+                                     (#'task.sync-databases/sync-and-analyze-database*! (u/the-id db))))))]
+            (testing "sense checks before deleting the database"
+              (testing "sense check 1: sync-and-analyze-database! should not log a warning"
+                (is (false? (cant-sync-logged?))))
+              (testing "sense check 2: triggering the sync via the POST /api/database/:id/sync_schema endpoint should succeed"
+                (is (= {:status "ok"}
+                       (mt/user-http-request :crowberto :post 200 (str "/database/" (u/the-id db) "/sync_schema"))))))
+            ;; release db resources like connection pools we don't have to wait to finish syncing before destroying the db
```
Comment (qnkhuat) [inline test/metabase/driver_test.clj:109]:
```suggestion
            ;; release db resources like connection pools so we don't have to wait to finish syncing before destroying the db
```

File: test/metabase/driver_test.clj [hunk 12]
```diff
@@ -68,24 +72,75 @@
...
+                          ;; but to an S3 bucket that may contain many databases
+                          ;; TODO: other drivers are still failing with this test. For these drivers there's a good chance there's a bug in
+                          ;; test.data.<driver> code, and not with the driver itself.
+                          (remove #{:athena :presto-jdbc :oracle :vertica}))
+      (let [database-name (mt/random-name)
+            dbdef         (basic-db-definition database-name)]
+        (mt/dataset dbdef
+          (let [db (mt/db)
+                cant-sync-logged? (fn []
+                                 (some?
```
Comment (qnkhuat) [inline test/metabase/driver_test.clj:95]: The indent here looks busted

File: src/metabase/task/sync_databases.clj [hunk 13]
```diff
@@ -72,6 +74,30 @@
...
+                  ;; purposes of creating a new H2 database.
+                  (binding [h2/*allow-testing-h2-connections* true]
+                    (driver.u/can-connect-with-details? (:engine database) (:details database) :throw-exceptions))
+                  nil
+                  (catch Throwable e
+                    e))]
+      (log/warnf ex "Cannot sync Database %s: %s" (:name database) (ex-message ex))
+      (do
+        (sync-metadata/sync-db-metadata! database)
+          ;; only run analysis if this is a "full sync" database
```
Comment (qnkhuat) [inline src/metabase/task/sync_databases.clj:95]:
```suggestion
        ;; only run analysis if this is a "full sync" database
```

File: modules/drivers/mongo/src/metabase/driver/mongo.clj [hunk 14]
```diff
@@ -55,10 +55,12 @@
 (defmethod driver/can-connect? :mongo
   [_ details]
   (with-mongo-connection [^DB conn, details]
-    (= (float (-> (cmd/db-stats conn)
-                  (m.conversion/from-db-object :keywordize)
-                  :ok))
-       1.0)))
+    (let [db-stats (-> (cmd/db-stats conn)
```
Comment (qnkhuat) [inline modules/drivers/mongo/src/metabase/driver/mongo.clj:58]: what is the reason behind this change?

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj [hunk 15]
```diff
@@ -225,13 +229,14 @@
...
+            (f *mongo-connection*))
+          (finally
+            (mg/disconnect mongo-client)
+            (log/debug (u/format-color 'cyan (trs "Closed MongoDB connection.")))))))))
 
 (defmacro with-mongo-connection
```
Comment (qnkhuat) [inline modules/drivers/mongo/src/metabase/driver/mongo/util.clj:241]: We need to update the docs to let ppl know *mongo-client* will be bound

**Comment by qnkhuat** [comment]
Given you fixed a handful of database destroy bugs, do you think you can add a test for it? it's not within the scope of this PR but would be nice to have it

File: test/metabase/driver_test.clj [hunk 16]
```diff
@@ -68,24 +72,75 @@
...
+                (is (= {:status "ok"}
+                       (mt/user-http-request :crowberto :post 200 (str "/database/" (u/the-id db) "/sync_schema"))))))
+            ;; release db resources like connection pools we don't have to wait to finish syncing before destroying the db
+            (driver/notify-database-updated driver/*driver* db)
+            ;; destroy the db
+            (if (contains? #{:redshift :snowflake} driver/*driver*)
+              ;; in the case of some cloud databases, the test database is never created, and shouldn't be destroyed.
+              ;; so fake it by redefining the database name on the dbdef
+              (t2/update! :model/Database (u/the-id db)
+                          {:details (assoc (:details (mt/db)) :db "fake-db-name-that-definitely-wont-be-used")})
```
Comment (qnkhuat) [inline test/metabase/driver_test.clj:116]: I think we can just do this update for all drivers, it'll make connecting to the DB fail anw.

Comment (calherries) [reply test/metabase/driver_test.clj:116]: The problem with that is this code depends on the key(s) in `details` that identifies the database, and the specifics of how the driver is tested. So we'd really prefer not to have to do it this way, and I wanted to avoid creating another driver multimethod for `metabase.test.data.interface` just for this test.

For example with redshift, we always use a single test DB, identified by the `:db` key ([here](https://github.com/metabase/metabase/blob/6360f1a5f6be43abfdcff6da98e9aac1c8416381/modules/drivers/redshift/test/metabase/test/data/redshift.clj#L45)). So to test what happens when a database is deleted, the test needs to know that `:db` is the key that identifies a DB for redshift.

**Comment by calherries** [comment]
> Given you fixed a handful of database destroy bugs, do you think you can add a test for it? it's not within the scope of this PR but would be nice to have it

Writing a test for each driver is possible but painful because there's no general way to tell a database exists other than `driver/can-connect?`, which can also have bugs as this PR shows.

Maybe the 80% solution we want is a separate, more focused test that only uses `tx/destroy-db!` and `driver/can-connect?`, while excluding all the sync stuff. It'll look very similar to this test but at least I can put it under `test/data`.

File: modules/drivers/mongo/src/metabase/driver/mongo.clj (hunk 14 above)
Comment (calherries) [reply modules/drivers/mongo/src/metabase/driver/mongo.clj:58]: I've just added that explanation as comments
https://github.com/metabase/metabase/blob/5dfee97b97e3234d6f605515a88b529716bd058a/modules/drivers/mongo/src/metabase/driver/mongo.clj#L61-L67

File: modules/drivers/sqlite/test/metabase/test/data/sqlite.clj [hunk 17]
```diff
@@ -1,16 +1,22 @@
...
             [metabase.test.data.sql-jdbc :as sql-jdbc.tx]
             [metabase.test.data.sql-jdbc.execute :as execute]))
 
+(set! *warn-on-reflection* true)
+
 (sql-jdbc.tx/add-test-extensions! :sqlite)
 
 (defmethod tx/supports-timestamptz-type? :sqlite [_driver] false)
 
+(defn- file-name [dbdef]
```
Comment (qnkhuat) [inline modules/drivers/sqlite/test/metabase/test/data/sqlite.clj:14]:
```suggestion
(defn- db-file-name [dbdef]
```

`file-name` alone doesn't make much sense.

**Review by qnkhuat — APPROVED** [review]
Cool, LGTM

File: modules/drivers/sqlite/test/metabase/test/data/sqlite.clj (hunk 17 above)
Comment (calherries) [reply modules/drivers/sqlite/test/metabase/test/data/sqlite.clj]: good point, thanks
